src/LabLog/Controllers/AdminController.cs

```diff
@@ -22,33 +23,48 @@ public AdminController(EventModelContext db)
 
         public IActionResult Index()
         {
-            ViewData["Rooms"]="";
-            return View();
+            List<RoomModel> roomList = new List<RoomModel>();
+            var events = from _events in _db.LabEvents
+                         where _events.EventType.Equals(RoomCreatedEvent.EventTypeString)
+                         select _events;
+            foreach (LabEvent roomCreatedEvent in events)
+            {
+                RoomModel room = new RoomModel();
+                room.ApplyRoomCreatedEvent(roomCreatedEvent);
+                roomList.Add(room);
+            }
+            return View(roomList);
         }
 
+        public IActionResult AddRoom()
+        {
+            return View();
+        }
 
         [HttpPost]
         [ValidateAntiForgeryToken]
-        public IActionResult Index(RoomModel room)
+        public IActionResult AddRoom(RoomModel room)

```

Comment:
I've moved the logic for form processing back into AddRoom, so that if there is a failure, the message can be displayed on the same page for the user to correct. However, I'm also trying to change the form action on the index.cshtml page to use `<form asp-controller="Admin" asp-action="AddRoom" method="post">` instead of `<form action="./admin/AddRoom">`.
This confuses it, because it doesn't know which AddRoom() route to go to. If I move the logic into a single AddRoom() route, then it automatically submits the empty form, because it thinks it has received data.
How can I test either where the page has come from (referrer), or otherwise determine whether AddRoom() has been arrived to for the first time, or after the user has tried to submit data?

Reply:
Sorted